### PR TITLE
Expose more CMake options for opentelemetry-cpp

### PR DESCRIPTION
**Pull request by sproberts92**
Specify library name and version:  **opentelemetry-cpp**

I was looking for some of these extra features that are hidden behind cmake options, and noticed that the recipe didn't expose them. So I've had a crack at exposing all the rest that are available in opentelemetry-cpp's CMakeLists.txt.

I've tried to keep the conan option defaults matching what the package was previously being built with, and for the new ones defaulting to what opentelemetry-cpp defaults them to. 

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by sproberts92** [comment]
@toge and @prince-chrismc, possibly interesting to both of you, as the two who reviewed my previous changes to this recipe a few months ago.

**Comment by sproberts92** [comment]
Looks like the failure is the result of Boost locale not supporting certain complier versions, which presumably we are using for the Mac OS builds: https://github.com/conan-io/conan-center-index/blob/master/recipes/boost/all/conanfile.py#L350-L370

Not sure how to proceed, trying to determine whether locale is definitely/still required.

**Comment by sproberts92** [comment]
Worked around the apple issue by applying the same condition that the boost recipe does, and dropping down to boost 1.8.0 in that case:

https://github.com/conan-io/conan-center-index/blob/master/recipes/boost/all/conanfile.py#L166

**Comment by prince-chrismc** [comment]
You need to update boost 
`WARN: thrift/0.17.0: requirement boost/1.81.0 overridden by opentelemetry-cpp/1.7.0 to boost/1.80.0`  :)

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 1]
```diff
@@ -53,19 +86,57 @@ def layout(self):
...
+                    requires.add("boost/1.80.0")
+                else:
+                    requires.add("boost/1.81.0")
+
+        if self.options.with_etw:
+            requires.add("nlohmann_json/3.11.2")
+            requires.add("openssl/1.1.1s")
+
+        for r in requires:
+            self.requires(r)
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:139]: Please just call self.requires :) we like to keep things minimal

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 2]
```diff
@@ -53,19 +86,57 @@ def layout(self):
...
 
     def validate(self):
         if self.info.settings.compiler.cppstd:
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:142]:
```suggestion
        if self.settings.get_safe("compiler.cppstd"):
```

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 3]
```diff
@@ -75,6 +146,12 @@ def validate(self):
         if self.settings.os != "Linux" and self.options.shared:
             raise ConanInvalidConfiguration(f"{self.ref} supports building shared libraries only on Linux")
 
+        if self.options.with_otlp_grpc and not self.options.with_otlp:
+            raise ConanInvalidConfiguration("Option 'with_otlp_grpc' requires 'with_otlp'")
+
+        if self.options.with_otlp_http and not self.options.with_otlp:
+            raise ConanInvalidConfiguration("Option 'with_otlp_http' requires 'with otlp'")
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:153]: These should be deleted in configure if the requirement option is false too

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 4]
```diff
@@ -53,19 +86,57 @@ def layout(self):
...
+                    requires.add("boost/1.81.0")
+
+        if self.options.with_etw:
+            requires.add("nlohmann_json/3.11.2")
+            requires.add("openssl/1.1.1s")
+
+        for r in requires:
+            self.requires(r)
 
     def validate(self):
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:141]: There should be a check here that boost::local is enabled

File: recipes/opentelemetry-cpp/all/conanfile.py (hunk 1 above)
Comment (sproberts92) [reply recipes/opentelemetry-cpp/all/conanfile.py:139]: I had that originally, the main reason I moved away from it is that there are several different exporters that require, for example libcurl. Specifying the same dependency multiple times resulting in the Conan 2 pipeline throwing `ConanException: Duplicated requirement` 

The boolean logic to figure out which circumstances to add a specific requirement seemed less minimal than using a set. But if that's preferred, I can change it back.

**Comment by sproberts92** [comment]
> You need to update boost `WARN: thrift/0.17.0: requirement boost/1.81.0 overridden by opentelemetry-cpp/1.7.0 to boost/1.80.0` :)

This is specifically for `apple-clang`, as boost 1.8.1 drops support for it, [see here](https://github.com/conan-io/conan-center-index/blob/master/recipes/boost/all/conanfile.py#L350-L370).

Is there a better way to achieve this?

File: recipes/opentelemetry-cpp/all/conanfile.py (hunk 1 above)
Comment (sproberts92) [reply recipes/opentelemetry-cpp/all/conanfile.py:139]: Ok, I managed to structure it in a way that isn't too bad and only add each one once.

File: recipes/opentelemetry-cpp/all/conanfile.py (hunk 2 above)
Comment (sproberts92) [reply recipes/opentelemetry-cpp/all/conanfile.py]: Done

File: recipes/opentelemetry-cpp/all/conanfile.py (hunk 3 above)
Comment (sproberts92) [reply recipes/opentelemetry-cpp/all/conanfile.py]: Done I think, if I understood correctly.

File: recipes/opentelemetry-cpp/all/conanfile.py (hunk 4 above)
Comment (sproberts92) [reply recipes/opentelemetry-cpp/all/conanfile.py:140]: I'm sorry, I don't quite understand this suggestion 😅

Comment (prince-chrismc) [reply recipes/opentelemetry-cpp/all/conanfile.py:140]: Sorry reviews when I am on my mobile are a little more self serve 🙊 here's what I was thinking... it a kinda common thing for boost dependencies

https://github.com/conan-io/conan-center-index/blob/b2beeacb7353aa3fc20a4e6d04039dab682559bf/recipes/gtsam/all/conanfile.py#L95-L105

Comment (prince-chrismc) [reply recipes/opentelemetry-cpp/all/conanfile.py:140]: This will block if the dep is not available in the current graph

Comment (sproberts92) [reply recipes/opentelemetry-cpp/all/conanfile.py:141]: Thanks, this approach is working.

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 5]
```diff
@@ -49,35 +82,88 @@ def configure(self):
             except Exception:
                 pass
 
+        if not self.options.with_otlp:
+            try:
+                del self.options.with_otlp_grpc
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:87]: This should use `rm_safe` :)

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 6]
```diff
@@ -49,35 +82,88 @@ def configure(self):
...
+            except Exception:
+                pass
+
+            try:
+                del self.options.with_otlp_http
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:92]: Rm safe here too

File: recipes/opentelemetry-cpp/all/conanfile.py (hunk 5 above)
Comment (prince-chrismc) [reply recipes/opentelemetry-cpp/all/conanfile.py:87]: required_conan_version should be 1.53 too

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 7]
```diff
@@ -49,35 +82,88 @@ def configure(self):
...
+            self.requires("abseil/20220623.0")
+
+        if self.options.with_otlp:
+            self.requires("protobuf/3.21.4")
+            if Version(self.version) <= "1.4.1":
+                self.requires("opentelemetry-proto/0.11.0")
+            else:
+                self.requires("opentelemetry-proto/0.19.0")
+
+            if self.options.with_otlp_grpc:
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:113]: This needs

```suggestion
            if self.options.get_safe("with_otlp_grpc"):
```

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 8]
```diff
@@ -49,35 +82,88 @@ def configure(self):
...
+                self.requires("grpc/1.50.1")
+
+        if (self.options.with_zipkin or
+           self.options.with_elasticsearch or
+           self.options.with_otlp_http or
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:118]: Get safe here too

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 9]
```diff
@@ -49,35 +82,88 @@ def configure(self):
...
+           self.options.with_etw):
+                self.requires("nlohmann_json/3.11.2")
+                self.requires("openssl/1.1.1s")
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:121]: Linter

```suggestion
             self.requires("openssl/1.1.1t")
```

File: recipes/opentelemetry-cpp/all/conanfile.py [hunk 10]
```diff
@@ -49,35 +82,88 @@ def configure(self):
...
     def validate(self):
-        if self.info.settings.compiler.cppstd:
+        if self.settings.get_safe("compiler.cppstd"):
             check_min_cppstd(self, self._minimum_cpp_standard)
         check_min_vs(self, "192")
 
         if self.settings.os != "Linux" and self.options.shared:
             raise ConanInvalidConfiguration(f"{self.ref} supports building shared libraries only on Linux")
 
+        if self.options.with_otlp_grpc and not self.options.with_otlp:
```
Comment (prince-chrismc) [inline recipes/opentelemetry-cpp/all/conanfile.py:149]: Since there are deleted its preferred yo use the get safe access

Same for the next one

**Review by prince-chrismc — COMMENTED** [review]
This looks really good, there's a hand full of access to the deleted options which should be gated with `get_safe` I did not mark all of them but  maybe a quick search and replace will do

**Comment by prince-chrismc** [comment]
Cla is stuck

File: recipes/opentelemetry-cpp/all/conanfile.py (hunk 5 above)
Comment (sproberts92) [reply recipes/opentelemetry-cpp/all/conanfile.py:87]: Ah I'm glad we now have a first-class way of doing this instead of the try/del/except!

**Review by prince-chrismc — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

**Pull request merged**